### PR TITLE
Download agent.

### PR DESCRIPTION
**Pull request by AlexanderVladimirov9090**
Done.

### CODE REVIEW

File: downloadagent/src/main/java/com/cloway/downloadagent/DownloadAgent.java [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+     */
+    public void download(String givenURL, String downloadDir) {
+
+        try {
+            URL uRL = new URL(givenURL);
+            URLConnection urlConnection = uRL.openConnection();
+            InputStream inputStream = urlConnection.getInputStream();
+            ReadableByteChannel channel = Channels.newChannel(inputStream);
+            ByteBuffer byteBuffer = ByteBuffer.allocate(64);
+            int size = urlConnection.getContentLength();
```
Comment (EmilGeorgiev) [inline downloadagent/src/main/java/com/cloway/downloadagent/DownloadAgent.java]: We can avoid loop when save in file by this way:
 
```java
URL website = new URL(givenURL);
ReadableByteChannel rbc = Channels.newChannel(website.openStream());
FileOutputStream fos = new FileOutputStream("information.html");
fos.getChannel().transferFrom(rbc, 0, Long.MAX_VALUE);
```

or use FileUtils from commons-io lib https://commons.apache.org/proper/commons-io/javadocs/api-2.5/org/apache/commons/io/FileUtils.html

```java
 URL url = new URL(givenURL)
 File file = new File(...) 
 FileUtils.copyURLToFile(url, file)
```

File: downloadagent/src/main/java/com/cloway/downloadagent/DownloadAgent.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    /**
+     * Writes to file.
+     *
+     * @param channel    readable byte channel for reading byte from recourse.
+     * @param byteBuffer
+     * @param size       size of file.
+     * @param fos        File output stream for writing in to file.
+     * @throws IOException when something is wrong with writing.
+     */
+    private void write(ReadableByteChannel channel, ByteBuffer byteBuffer, int size, FileOutputStream fos) throws IOException {
```
Comment (EmilGeorgiev) [inline downloadagent/src/main/java/com/cloway/downloadagent/DownloadAgent.java]: this is unnecessatry

File: downloadagent/src/test/java/DownloadFileFromResource.java [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+ * Created by clouway on 12.09.16.
+ *
+ * @author Alexander Vladimirov
+ *         <alexandervladimirov1902@gmail.com>
+ */
+public class DownloadFileFromResource {
+
+    @Test
+    public void happyPath() {
+        String urlString = "http://www.spyderonlines.com/images/wallpapers/picture/picture-8.jpg";
```
Comment (EmilGeorgiev) [inline downloadagent/src/test/java/DownloadFileFromResource.java]: if you do not have internet tests will fail

File: downloadagent/src/test/java/DownloadFileFromResource.java [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+        String downloadDir = "/home/clouway/Downloads";
+        ProgressBar progressBar = new RealProgressBar();
+        DownloadAgent downloadAgent = new DownloadAgent(progressBar);
+
+        downloadAgent.download(urlString, downloadDir);
+
+        File file = new File("/home/clouway/Downloads/picture-8.jpg");
+        assertThat(file.exists(), is(true));
```
Comment (EmilGeorgiev) [inline downloadagent/src/test/java/DownloadFileFromResource.java]: Задачата трябва да се тества автоматизирано и изолирано от всякакъв environment (дори и интернет) с юнит тестове. Трябва да се провери както цялостта на смъкнатите файлове, така и начина по - който прогреса се обновява.

File: downloadagent/src/main/java/com/cloway/downloadagent/DownloadAgent.java (hunk 1 above)
Comment (AlexanderVladimirov9090) [inline downloadagent/src/main/java/com/cloway/downloadagent/DownloadAgent.java]: I'm aware of this two approaches but, can't think of a way to give data needed for the progressBar.

File: downloadagent/src/test/java/DownloadFileFromResource.java (hunk 3 above)
Comment (AlexanderVladimirov9090) [inline downloadagent/src/test/java/DownloadFileFromResource.java]: yeah will use file:// scheme.

**Comment by AlexanderVladimirov9090** [comment]
Done.

File: downloadagent/src/main/java/com/cloway/downloadagent/DownloadAgent.java (hunk 1 above)
Comment (AlexanderVladimirov9090) [inline downloadagent/src/main/java/com/cloway/downloadagent/DownloadAgent.java]: Done and this one.

**Comment by EmilGeorgiev** [comment]
Is this PR ready for review ?

**Comment by AlexanderVladimirov9090** [comment]
I`m working on it now.

**Comment by AlexanderVladimirov9090** [comment]
Ready for review.

**Comment by AlexanderVladimirov9090** [comment]
Done with DownloadAgent.

File: downloadagent/src/test/java/com/cloway/downloadagent/DownloadFileFromResource.java [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+    @Test
+    public void happyPath(){
+        String urlString = "file:///home/clouway/workspaces/networking-and-gui/downloadagent/expected/picture-11.jpg";
+        String downloadDir = "/home/clouway/workspaces/networking-and-gui/downloadagent/actual";
+        DownloadAgent downloadAgent = new DownloadAgent(progressBar);
+
+        downloadAgent.download(urlString, downloadDir);
+        File file = new File("/home/clouway/workspaces/networking-and-gui/downloadagent/actual/picture-11.jpg");
+
+        assertThat(file.isFile(), is(true));
```
Comment (EmilGeorgiev) [inline downloadagent/src/test/java/com/cloway/downloadagent/DownloadFileFromResource.java]: We should assert bytes. By this way we will guarantees that two images are equals

Comment (AlexanderVladimirov9090) [inline downloadagent/src/test/java/com/cloway/downloadagent/DownloadFileFromResource.java]: Will be done.

Comment (AlexanderVladimirov9090) [inline downloadagent/src/test/java/com/cloway/downloadagent/DownloadFileFromResource.java]: And will try something for better testing.

**Comment by AlexanderVladimirov9090** [comment]
Got The test Right this time.

**Comment by AlexanderVladimirov9090** [comment]
And I`m done with Download Agent.

**Comment by EmilGeorgiev** [comment]
LGTM

**Pull request merged**